File: server/app/schemas/session_record.py
"""
Session record schemas — read-only views of historical conversations
"""
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.satisfaction_survey_record import SatisfactionSummaryResponse
from app.schemas.ticket import RelatedTicketResponse
from app.schemas.queue import QueueRecordBrief


class SessionRecordVisitor(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    public_id: str
    external_id: str
    name: str
    avatar_color: str | None = None


class SessionRecordAgent(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    display_name: str | None = None
    avatar: str | None = None


class SessionRecordChannel(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    channel_type: str


class ReceptionParticipant(BaseModel):
    agent_id: int
    name: str | None = None


class SessionRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    public_id: str
    share_code: str
    session_type: str | None = None
    bot_handoff_status: str | None = None
    visitor: SessionRecordVisitor | None = None
    agent: SessionRecordAgent | None = None
    channel: SessionRecordChannel | None = None
    status: str
    started_at: datetime | None = None
    ended_at: datetime | None = None
    ended_by: str | None = None
    duration_seconds: int | None = None
    visitor_system: str | None = None
    visitor_browser: str | None = None
    visitor_ip: str | None = None
    created_at: datetime | None = None
    message_count: int = 0
    visitor_message_count: int = 0
    agent_message_count: int = 0
    bot_phase_message_count: int = 0
    human_phase_message_count: int = 0
    human_phase_visitor_message_count: int = 0
    human_phase_agent_message_count: int = 0
    satisfaction: SatisfactionSummaryResponse | None = None
    last_assigned_queue: QueueRecordBrief | None = None
    queue_duration_seconds: int | None = None
    first_human_response_seconds: int | None = None
    agent_response_count: int | None = None
    agent_avg_response_seconds: int | None = None
    has_queue: bool = False
    queue_entered_at: datetime | None = None
    queue_assigned_at: datetime | None = None
    queue_result: str | None = None
    reception_segment_count: int = 0
    reception_transfer_count: int = 0
    reception_final_agent_id: int | None = None
    reception_participants: list[ReceptionParticipant] = Field(default_factory=list)
    reception_generation_status: str | None = None


class SessionRecordListResponse(BaseModel):
    items: list[SessionRecordResponse]
    total: int
    page: int
    per_page: int
    pages: int


class SessionRecordDetailResponse(SessionRecordResponse):
    """Extended detail with group info."""
    last_message_preview: str | None = None
    related_tickets: list[RelatedTicketResponse] = Field(default_factory=list)


class SessionRecordMessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    conversation_id: int
    sender_type: str
    sender_id: int | None = None
    sender_name: str | None = None
    sender_avatar: str | None = None
    content_type: str
    content: str
    metadata: dict = Field(default_factory=dict)
    created_at: datetime
    event_type: str | None = None
    satisfaction_record_id: int | None = None
    config_version: int | None = None


class SessionRecordMessageListResponse(BaseModel):
    items: list[SessionRecordMessageResponse]
    has_more: bool


class SessionRecordSegmentResponse(BaseModel):
    seq_no: int
    agent_id: int | None = None
    agent_name: str | None = None
    group_id: int | None = None
    group_name: str | None = None
    started_at: datetime
    ended_at: datetime | None = None
    duration_seconds: int | None = None
    entry_reason: str
    end_reason: str | None = None
    from_agent_id: int | None = None
    to_agent_id: int | None = None
    visitor_message_count: int = 0
    agent_message_count: int = 0
    first_response_seconds: int | None = None
    avg_response_seconds: int | None = None


class ReceptionTrajectoryResponse(BaseModel):
    conversation_id: int
    conversation_status: str
    generation_status: str | None = None
    segments: list[SessionRecordSegmentResponse] = Field(default_factory=list)

File: web/models/session-record.ts
import type { SatisfactionSummary } from './satisfaction-survey'
import type { RelatedTicket } from './ticket'

export type SessionRecordVisitor = {
  id: number
  public_id: string
  external_id: string
  name: string
  avatar_color: string | null
}

export type SessionRecordAgent = {
  id: number
  name: string
  display_name: string | null
  avatar: string | null
}

export type SessionRecordChannel = {
  id: number
  name: string
  channel_type: string
}

export type QueueRecordBrief = {
  queue_type: 'employee_group' | 'employee' | string
  queue_id: number
  name: string
}

export type SessionRecordType = 'human' | 'bot' | 'bot_human'

export type BotHandoffStatus =
  | 'not_triggered'
  | 'waiting_confirmation'
  | 'handoff_in_progress'
  | 'in_queue'
  | 'succeeded'
  | 'failed'
  | 'dismissed'

export type QueueResult =
  | 'assigned'
  | 'canceled'
  | 'timeout'
  | 'waiting'
  | 'failed'

export type ReceptionParticipant = {
  agent_id: number
  name: string | null
}

export type ReceptionGenerationStatus = 'generated' | 'failed' | null

export type ReceptionSegment = {
  seq_no: number
  agent_id: number | null
  agent_name: string | null
  group_id: number | null
  group_name: string | null
  started_at: string
  ended_at: string | null
  duration_seconds: number | null
  entry_reason: string
  end_reason: string | null
  from_agent_id: number | null
  to_agent_id: number | null
  visitor_message_count: number
  agent_message_count: number
  first_response_seconds: number | null
  avg_response_seconds: number | null
}

export type ReceptionTrajectory = {
  conversation_id: number
  conversation_status: SessionRecord['status']
  generation_status: ReceptionGenerationStatus
  segments: ReceptionSegment[]
}

export type SessionRecord = {
  id: number
  public_id: string
  share_code: string
  session_type: SessionRecordType | null
  bot_handoff_status: BotHandoffStatus | null
  visitor: SessionRecordVisitor | null
  agent: SessionRecordAgent | null
  channel: SessionRecordChannel | null
  status: 'queued' | 'active' | 'bot' | 'handoff_pending' | 'closed'
  started_at: string | null
  ended_at: string | null
  ended_by: string | null
  duration_seconds: number | null
  visitor_system: string | null
  visitor_browser: string | null
  visitor_ip: string | null
  created_at: string | null
  message_count: number
  visitor_message_count: number
  agent_message_count: number
  bot_phase_message_count: number
  human_phase_message_count: number
  human_phase_visitor_message_count: number
  human_phase_agent_message_count: number
  satisfaction: SatisfactionSummary | null
  last_assigned_queue: QueueRecordBrief | null
  queue_duration_seconds: number | null
  first_human_response_seconds: number | null
  agent_response_count: number | null
  agent_avg_response_seconds: number | null
  has_queue: boolean
  queue_entered_at: string | null
  queue_assigned_at: string | null
  queue_result: QueueResult | null
  reception_segment_count: number
  reception_transfer_count: number
  reception_final_agent_id: number | null
  reception_participants: ReceptionParticipant[]
  reception_generation_status: ReceptionGenerationStatus
}

export type SessionRecordListResponse = {
  items: SessionRecord[]
  total: number
  page: number
  per_page: number
  pages: number
}

export type SessionRecordDetail = SessionRecord & {
  last_message_preview: string | null
  related_tickets: RelatedTicket[]
}

export type SessionRecordMessage = {
  id: number
  conversation_id: number
  sender_type: 'visitor' | 'agent' | 'bot' | 'system'
  sender_id: number | null
  sender_name: string | null
  sender_avatar: string | null
  content_type: 'text' | 'rich_text' | 'image' | 'file' | 'system' | 'welcome' | 'bot_welcome' | 'satisfaction_event' | 'internal_note'
  content: string
  metadata?: Record<string, unknown>
  created_at: string
  event_type?: string
  satisfaction_record_id?: number
  config_version?: number
}

export type SessionRecordMessageListResponse = {
  items: SessionRecordMessage[]
  has_more: boolean
}

export type SessionRecordFilters = {
  page: number
  per_page: number
  start_date?: string
  end_date?: string
  agent_id?: number
  visitor_id?: number
  session_type?: SessionRecordType
  has_queue?: boolean
  keyword?: string
  satisfaction_status?: string
  satisfaction_resolved?: string
  satisfaction_service_option?: string
  satisfaction_service_label?: string
  satisfaction_product_option?: string
  satisfaction_product_label?: string
}
